### PR TITLE
Add sr-iov utils package

### PR DESCRIPTION
**Pull request by lvfxx**

### CODE REVIEW

File: pkg/sriov/utils/utils_provider.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+func (u *sriovUtilsProvider) GetSriovVirtualFunctionsCapacity(pfPciAddr string) (int, error) {
+	totalVfFilePath := filepath.Join(u.sysfsDevicesPath, pfPciAddr, totalVfFile)
+	vfs, err := ioutil.ReadFile(filepath.Clean(totalVfFilePath))
+	if err != nil {
+		return 0, errors.Errorf("unable to locate sriov_totalvfs file for device %s: %v", pfPciAddr, err)
+	}
+	totalVfs := bytes.TrimSpace(vfs)
+	numTotalVfs, err := strconv.Atoi(string(totalVfs))
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:132]: minor
```suggestion
	totalVfs := strings.TrimSpace(string(vfsBytes))
	numTotalVfs, err := strconv.Atoi(totalVfs)
```

File: pkg/sriov/utils/utils_provider.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+	if err != nil {
+		return 0, errors.Errorf("unable to convert string to int from sriov_totalvfs file for device %s: %v", pfPciAddr, err)
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:134]:
```suggestion
		return 0, errors.Errorf("unable to convert string to int from sriov_totalvfs file for device %s: %v", pfPciAddr, err.Error())
```

File: pkg/sriov/utils/utils_provider.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+func (u *sriovUtilsProvider) GetNetInterfacesNames(pciAddr string) ([]string, error) {
+	var names []string
+	netDir := filepath.Join(u.sysfsDevicesPath, pciAddr, netInterfacesPath)
+	if _, err := os.Lstat(netDir); err != nil {
+		return nil, errors.Errorf("no net directory under pci device %s: %v", pciAddr, err)
+	}
+
+	fInfos, err := ioutil.ReadDir(netDir)
+	if err != nil {
+		return nil, errors.Errorf("failed to read net directory %s: %v", netDir, err)
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:167]: Let's use `errors.Wrapf` here

File: pkg/sriov/utils/utils_provider.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+		devicePath := filepath.Join(u.sysfsDevicesPath, pciAddr)
+		return u.isFileOrLinkExists(devicePath), nil
+	}
+	return false, errors.Errorf("invalid pci address provided: %s", pciAddr)
+}
+
+func (u *sriovUtilsProvider) GetNetInterfacesNames(pciAddr string) ([]string, error) {
+	var names []string
+	netDir := filepath.Join(u.sysfsDevicesPath, pciAddr, netInterfacesPath)
+	if _, err := os.Lstat(netDir); err != nil {
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:161]: isFileOrLinkExists?

File: pkg/sriov/utils/utils_provider.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+func (u *sriovUtilsProvider) CreateVirtualFunctions(pfPciAddr string, vfNumber int) error {
+	if vfNumber < 1 {
+		return errors.Errorf("invalid number of virtual functions specified: %d. Must be > 0", vfNumber)
+	}
+	if u.IsSriovConfigured(pfPciAddr) {
+		return errors.Errorf("virtual functions are already exist for device: %s", pfPciAddr)
+	}
+	configuredVfPath := filepath.Join(u.sysfsDevicesPath, pfPciAddr, configuredVfFile)
+	err := ioutil.WriteFile(configuredVfPath, []byte(strconv.FormatInt(int64(vfNumber), 10)), 0600)
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:185]: Can we use this constat here?
```suggestion
	err := ioutil.WriteFile(configuredVfPath, []byte(strconv.FormatInt(int64(vfNumber), 10)), os.ModePerm)
```

File: pkg/sriov/utils/utils_provider.go [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+	if err != nil {
+		return errors.Errorf("failed to write virtual functions number for device %s: %v", pfPciAddr, err)
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:187]: Wrapf

File: pkg/sriov/utils/utils_provider.go [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+	}
+	return nil
+}
+
+func (u *sriovUtilsProvider) GetVirtualFunctionsList(pfPciAddr string) (vfList []string, err error) {
+	vfList = make([]string, 0)
+	pfDir := filepath.Join(u.sysfsDevicesPath, pfPciAddr)
+	_, err = os.Lstat(pfDir)
+	if err != nil {
+		err = errors.Errorf("could not get PF directory information for device %s: %v", pfPciAddr, err)
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:197]: Wrapf

File: pkg/sriov/utils/utils_provider.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+	configuredVfPath := filepath.Join(u.sysfsDevicesPath, pfPciAddr, configuredVfFile)
+	err := ioutil.WriteFile(configuredVfPath, []byte(strconv.FormatInt(int64(vfNumber), 10)), 0600)
+	if err != nil {
+		return errors.Errorf("failed to write virtual functions number for device %s: %v", pfPciAddr, err)
+	}
+	return nil
+}
+
+func (u *sriovUtilsProvider) GetVirtualFunctionsList(pfPciAddr string) (vfList []string, err error) {
+	vfList = make([]string, 0)
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:193]: Note: go allow to append to nil slices 
```suggestion
```

File: pkg/sriov/utils/utils_provider.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	pfDir := filepath.Join(u.sysfsDevicesPath, pfPciAddr)
+	_, err = os.Lstat(pfDir)
+	if err != nil {
+		err = errors.Errorf("could not get PF directory information for device %s: %v", pfPciAddr, err)
+		return
+	}
+
+	vfDirs, err := filepath.Glob(filepath.Join(pfDir, "virtfn*"))
+	if err != nil {
+		err = errors.Errorf("error reading VF directories %v", err)
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:203]: Wrapf

File: pkg/sriov/utils/utils_provider.go [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+	if err != nil {
+		return errors.Errorf("failed to write virtual functions number for device %s: %v", pfPciAddr, err)
+	}
+	return nil
+}
+
+func (u *sriovUtilsProvider) GetVirtualFunctionsList(pfPciAddr string) (vfList []string, err error) {
+	vfList = make([]string, 0)
+	pfDir := filepath.Join(u.sysfsDevicesPath, pfPciAddr)
+	_, err = os.Lstat(pfDir)
```
Comment (denis-tingajkin) [inline pkg/sriov/utils/utils_provider.go:195]: isFileOrLinkExists?

File: pkg/sriov/utils/utils_provider.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+	return
+}
+
+func (u *sriovUtilsProvider) isFileExists(filePath string) bool {
+	_, err := os.Stat(filePath)
+	return err == nil
+}
+
+func (u *sriovUtilsProvider) isFileOrLinkExists(filePath string) bool {
+	_, err := os.Lstat(filePath)
```
[the remark this answers is not in the document]
Comment (lvfxx) [reply pkg/sriov/utils/utils_provider.go:227]: We can use `isExists()` it this form:
```go
_, err := os.Stat(filePath)
if err != nil {
	return os.IsExist(err)
}
return true
```

File: pkg/sriov/utils/utils_provider.go (hunk 8 above)
Comment (lvfxx) [reply pkg/sriov/utils/utils_provider.go:193]: Didn't know it, thanks

File: pkg/sriov/utils/utils_provider.go (hunk 2 above)
Comment (lvfxx) [reply pkg/sriov/utils/utils_provider.go:134]: Decided to use `errors.Wrapf` here

File: pkg/sriov/utils/utils_provider.go (hunk 5 above)
Comment (lvfxx) [reply pkg/sriov/utils/utils_provider.go:185]: `os.ModePerm` = 0777. When i write 0777 by myself without this os constant, golangci-lint complains:
> G306: Expect WriteFile permissions to be 0600 or less (gosec)

But when i use `os.ModePerm` golangci-lint says it's OK. Looks like linter isn't very smart to detect that it is the same 0777 is complains about.
So, i think we should use less permissive bits.

File: pkg/sriov/utils/utils_provider.go [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+)
+
+var (
+	initRegexpOnce    sync.Once
+	validLongPCIAddr  *regexp.Regexp
+	validShortPCIAddr *regexp.Regexp
+)
+
+// SriovUtilsProvider provides utility methods for SR-IOV machinery
+type SriovUtilsProvider interface {
```
[the remark this answers is not in the document]
Comment (lvfxx) [reply pkg/sriov/utils/utils_provider.go:47]: Removed it

**Pull request merged**